src/mgr/ClusterState.h

```diff
@@ -127,6 +127,15 @@ class ClusterState
     return objecter->with_osdmap(std::forward<Args>(args)...);
   }
 
+  template<typename... Args>
+  auto with_pgmap_and_osdmap(Args &&... args) const ->
+    decltype(objecter->with_osdmap(const PGMap& pg_map, std::forward<Args>(args)...))
+  {
+    ceph_assert(objecter != nullptr);
+    std::lock_guard l(lock);
+    return objecter->with_osdmap(pg_map, std::forward<Args>(args)...);
+  }
+
```

Comment:
this needs to take a function object as the first argument. objecter->with_osdmap() will pass its osdmap as the first argument to that callback, so pg_map will be the second argument. i renamed to with_osdmap_and_pgmap() to reflect that order:
```suggestion
  // call cb(osdmap, pg_map, ...args) with the appropriate locks
  template <typename Callback, typename ...Args>
  auto with_osdmap_and_pgmap(Callback&& cb, Args&& ...args) const {
    ceph_assert(objecter != nullptr);
    std::lock_guard l(lock);
    return objecter->with_osdmap(std::forward<Callback>(cb), pg_map, std::forward<Args>(args)...);
  }
```